apollo-federation/src/sources/connect/json_selection/pretty.rs

```diff
@@ -0,0 +1,377 @@
+//! Pretty printing utility methods
+//!
+//! Working with raw JSONSelections when doing snapshot testing is difficult to
+//! read and makes the snapshots themselves quite large. This module adds a new
+//! pretty printing trait which is then implemented on the various sub types
+//! of the JSONSelection tree.
+
+use std::fmt::Write;
+
+use crate::sources::connect::json_selection::Alias;
+use crate::sources::connect::json_selection::JSONSelection;
+use crate::sources::connect::json_selection::NamedSelection;
+use crate::sources::connect::json_selection::PathSelection;
+use crate::sources::connect::json_selection::StarSelection;
+use crate::sources::connect::json_selection::SubSelection;
+
+/// Pretty print trait
+///
+/// This trait marks a type as supporting pretty printing itself outside of a
+/// Display implementation, which might be more useful for snapshots.
+pub trait PrettyPrintable {
+    /// Pretty print the struct
+    fn pretty_print(&self) -> Result<String, std::fmt::Error> {
```

Comment:
The main motivation was the fact that using `write!` allows for directly using formatted strings vs having to use the nested `.push_str(&format!(...))`. If you get the chance to simplify this in your next PR, please do!

Reply:
you should definitely use `write!()` instead of `push_str(&format!())`. If the output is known to be a string, `write!()` is infallible, so you can do `write!().unwrap()` or even `_ = write!()` to silence the warning.

Reply:
That said, even better would be if the pretty printing trait used an `f: &mut fmt::Formatter` parameter for output, to avoid allocating many intermediate strings. That is not something that needs to be done in this PR of course :)

Reply:
To clarify, I wasn't suggesting `push_str(&format!(...))` but just `push_str(&str)` (and sometimes `push(char)`), as in the following diff:
```diff
@@ -378,26 +389,29 @@ impl PathSelection {
         }
     }
 
-    pub fn pretty_print_with_indentation(&self, indentation: usize) -> Result<String, std::fmt::Error> {
+    pub fn pretty_print_with_indentation(&self, indentation: usize) -> String {
         let mut result = String::new();
 
         match self {
             PathSelection::Var(var, path) => {
-                let rest = path.pretty_print(indentation)?;
-                write!(result, "${var}{rest}")?;
+                let path_string = path.pretty_print(indentation);
+                result.push_str(var.as_str());
+                result.push_str(path_string.as_str());
             }
             PathSelection::Key(key, path) => {
-                let rest = path.pretty_print(indentation)?;
-                write!(result, "{key}{rest}")?;
+                let path_string = path.pretty_print(indentation);
+                result.push_str(key.dotted().as_str());
+                result.push_str(path_string.as_str());
             }
             PathSelection::Selection(sub) => {
-                let sub = sub.pretty_print(indentation)?;
-                write!(result, " {sub}")?;
+                let sub_string = sub.pretty_print(indentation);
+                result.push(' ');
+                result.push_str(sub_string.as_str());
             }
             PathSelection::Empty => {}
-        }
+        };
 
-        Ok(result)
+        result
     }
 }
```